Dockerfile

```diff
@@ -1,24 +1,102 @@
-FROM ruby:2.6.3-alpine
+#### Common Build Image Stage ####
 
-ENV APP_HOME /app
+FROM ruby:2.6.3-alpine AS common-build-env
 
+ARG APP_HOME=/app
+ARG BUILD_PACKAGES="build-base"
+ARG DEV_PACKAGES="postgresql-dev git nodejs yarn"
+ARG RUBY_PACKAGES="tzdata"
+ARG bundleWithout=""
+
+ENV BUNDLER_VERSION="2.0.2"
+ENV BUNDLE_PATH="/gems"
+ENV BUNDLE_WITHOUT=${bundleWithout}
+
+RUN apk update && \
+    apk upgrade && \
+    apk add --update --no-cache $BUILD_PACKAGES $DEV_PACKAGES $RUBY_PACKAGES && \
+    gem update --system && \
+    find / -wholename '*default/bundler-*.gemspec' -delete && \
+    rm /usr/local/bin/bundle && \
+    rm /usr/local/bin/bundler && \
+    gem install bundler -v $BUNDLER_VERSION
+
+WORKDIR $APP_HOME
+    
+COPY Gemfile Gemfile.lock package.json yarn.lock ./
+
+RUN bundle install && \
+    yarn install
+
+
+#### Dev build builds on common build env image ####
+
+FROM common-build-env AS dev-build
+ARG APP_HOME=/app
+ARG railsEnv="development"
+ENV RAILS_ENV=${railsEnv}
+
+WORKDIR $APP_HOME
+
+COPY . .
+
+CMD bundle exec rails db:migrate && bundle exec rails server -b 0.0.0.0
+
+
+#### Prod minify stage builds the app on the common build env image ####
+
+FROM common-build-env AS prod-minify
+
+ARG APP_HOME=/app
+ARG railsEnv="production"
+
+ENV RAILS_ENV=${railsEnv}
 # These variables are required for running Rails processes like assets:precompile
 ENV GOVUK_NOTIFY_API_KEY=TestKey
 ENV DOMAIN=dummy.build.domain
-
-RUN apk add --update build-base postgresql-dev git tzdata nodejs yarn && \
-    cp /usr/share/zoneinfo/Europe/London /etc/localtime && \
-    echo "Europe/London" > /etc/timezone
+ENV STAGING_DOMAIN=dummy.build.domain
+ENV SECRET_KEY_BASE=TestKey
 
 WORKDIR $APP_HOME
-COPY Gemfile Gemfile.lock package.json yarn.lock $APP_HOME/
 
-RUN gem install bundler:2.0.2 && \
-    yarn install --check-files && \
-    bundle install
+RUN rm -rf $BUNDLE_PATH/cache/*.gem && \
+    find $BUNDLE_PATH/gems/ -name "*.c" -delete && \
+    find $BUNDLE_PATH/gems/ -name "*.o" -delete
 
-COPY . $APP_HOME/
+COPY . .
 
 RUN bundle exec rake assets:precompile
 
+RUN rm -rf tmp/cache app/assets lib/assets vendor/assets node_modules
+
+
+#### Production image builds from scratch and copies the build app components from the prod minify stage ####
+
+FROM ruby:2.6.3-alpine AS prod-build
+ARG railsEnv="production"
+ARG bundleWithout=""
+ARG APP_HOME=/app
+ARG PACKAGES="tzdata postgresql-client"
+
+ENV RAILS_ENV=${railsEnv}
+ENV BUNDLE_WITHOUT=${bundleWithout}
+ENV BUNDLE_PATH="/gems"
```

Comment:
Again, as this depends on prior images, can we get away with declaring it once at the top?

Reply:
Because of the repeated use of FROM we can't :(

Reply:
gah